pkg/file/joiner/joiner.go

```diff
@@ -210,6 +210,55 @@ func (j *joiner) Seek(offset int64, whence int) (int64, error) {
 
 }
 
+func (j *joiner) IterateChunkAddresses(fn swarm.AddressIterFunc) error {
+	// report root address
+	stop := fn(j.addr)
+	if stop {
+		return nil
+	}
+
+	var eg errgroup.Group
+	j.processChunkAddresses(fn, j.rootData, j.span, &eg)
+
+	return eg.Wait()
+}
+
+func (j *joiner) processChunkAddresses(fn swarm.AddressIterFunc, data []byte, subTrieSize int64, eg *errgroup.Group) {
+	// we are at a leaf data chunk
+	if subTrieSize <= int64(len(data)) {
+		return
+	}
+
+	for cursor := 0; cursor < len(data); cursor += j.refLength {
+		select {
+		case <-j.ctx.Done():
+			return
+		default:
+		}
+
+		address := swarm.NewAddress(data[cursor : cursor+j.refLength])
+
+		stop := fn(address)
+		if stop {
+			break
+		}
+
+		func(address swarm.Address, eg *errgroup.Group) {
+			eg.Go(func() error {
+				ch, err := j.getter.Get(j.ctx, storage.ModeGetRequest, address)
```

Comment:
at a second look, since you first `Get` but only decide if to quit the recursive call when the subtrie size is `<= len(data)` then it means you have already retrieved the data chunk itself